template/tree_v4.go

```diff
@@ -453,91 +466,57 @@ func (t *TreeV4) Delete(address patricia.IPv4Address, matchFunc MatchesFunc, mat
 	targetNode.Left = 0
 	targetNode.Right = 0
 	t.availableIndexes = append(t.availableIndexes, targetNodeIndex)
-	return deleteCount, nil
+	return deleteCount
 }
 
 // FindTagsWithFilter finds all matching tags that passes the filter function
-func (t *TreeV4) FindTagsWithFilter(address patricia.IPv4Address, filterFunc FilterFunc) ([]GeneratedType, error) {
-	root := &t.nodes[1]
-	if filterFunc == nil {
-		return t.FindTags(address)
-	}
-
-	var matchCount uint
+// - use FindTagsWithFilterAppend if you can reuse slices, to cut down on allocations
+func (t *TreeV4) FindTagsWithFilter(address patricia.IPv4Address, filterFunc FilterFunc) []GeneratedType {
 	ret := make([]GeneratedType, 0)
+	return t.FindTagsWithFilterAppend(ret, address, filterFunc)
+}
 
-	if root.TagCount > 0 {
-		for _, tag := range t.tagsForNode(1) {
-			if filterFunc(tag) {
-				ret = append(ret, tag)
-			}
-		}
-	}
+// FindTagsWithFilterAppend finds all matching tags that passes the filter function
+// - results are appended to the input slice
+func (t *TreeV4) FindTagsWithFilterAppend(ret []GeneratedType, address patricia.IPv4Address, filterFunc FilterFunc) []GeneratedType {
+	retPos := len(ret)
+	ret = t.FindTagsAppend(ret, address)
 
-	if address.Length == 0 {
-		// caller just looking for root tags
-		return ret, nil
+	if len(ret) == retPos || filterFunc == nil {
+		return ret
 	}
 
-	var nodeIndex uint
-	if !address.IsLeftBitSet() {
-		nodeIndex = root.Left
-	} else {
-		nodeIndex = root.Right
-	}
-
-	// traverse the tree
-	count := 0
-	for {
-		count++
-		if nodeIndex == 0 {
-			return ret, nil
-		}
-		node := &t.nodes[nodeIndex]
-
-		matchCount = node.MatchCount(address)
-		if matchCount < node.prefixLength {
-			// didn't match the entire node - we're done
-			return ret, nil
-		}
-
-		// matched the full node - get its tags, then chop off the bits we've already matched and continue
-		if node.TagCount > 0 {
-			for _, tag := range t.tagsForNode(nodeIndex) {
-				if filterFunc(tag) {
-					ret = append(ret, tag)
-				}
-			}
-		}
-
-		if matchCount == address.Length {
-			// exact match - we're done
-			return ret, nil
-		}
-
-		// there's still more address - keep traversing
-		address.ShiftLeft(matchCount)
-		if !address.IsLeftBitSet() {
-			nodeIndex = node.Left
-		} else {
-			nodeIndex = node.Right
+	// filter in place
```

Comment:
This still allocates less than before, where a new slice was allocated for the pre-filtered results, then thrown away after the filtered results were accepted. If you're reusing the same buffer over and over, the allocations will happen once, and then reused. But yeah, I think it's worth running the filter func earlier in the stack. I'll take a look, thanks.

Reply:
Updated in https://github.com/kentik/patricia/pull/14/commits/94f2b4b98276939729c217ed336a8bd22adec731

Reply:
Yeah, my main concern was if by passing around the filter function and doing these extra checks it would slow things down. Based on offline discussion it doesn't seem like that is the case, and the new changes look good to me.